lib/xgu/xgu.h

```diff
@@ -0,0 +1,937 @@
+}
+
+inline
+uint32_t* xgu_set_light_diffuse_color(uint32_t* p, unsigned int light_index, uint32_t color) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    return push_command_parameter(p, NV097_SET_LIGHT_DIFFUSE_COLOR + light_index*4, color);
+}
+
+inline
+uint32_t* xgu_set_light_specular_color(uint32_t* p, unsigned int light_index, uint32_t color) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    return push_command_parameter(p, NV097_SET_LIGHT_SPECULAR_COLOR + light_index*4, color);
+}
+
+inline
+uint32_t* xgu_set_light_local_range(uint32_t* p, unsigned int light_index, float range) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    return push_command_float(p, NV097_SET_LIGHT_LOCAL_RANGE + light_index*4, range);
+}
+
+inline
+uint32_t* xgu_set_light_infinite_half_vector(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_INFINITE_HALF_VECTOR + light_index*3*4, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_infinite_direction(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_INFINITE_DIRECTION + light_index*3*4, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_spot_falloff(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_SPOT_FALLOFF + light_index*3*4, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_spot_direction(uint32_t* p, unsigned int light_index, XguVec4 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_SPOT_DIRECTION + light_index*4*4, 4);
+    return push_floats(p, v.f, 4);
+}
+
+inline
+uint32_t* xgu_set_light_local_position(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_LOCAL_POSITION + light_index*3*4, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_local_attenuation(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index >= 0);
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_LOCAL_ATTENUATION + light_index*3*4, 3);
+    return push_floats(p, v.f, 3);
+}
+
+/* ==== Texture stuff ==== */
+
+inline
+uint32_t* xgu_set_texture_offset(uint32_t* p, uint32_t offset) {
+    return push_command_parameter(p, NV097_SET_TEXTURE_OFFSET, offset);
+}
+
+inline
+uint32_t* xgu_set_texture_format(uint32_t* p, uint8_t context_dma, bool cubemap_enable,
+                                 XguBorderSrc border_src, uint8_t dimensionality,
```

Comment:
[OpenGL 1.x / 2.x specifies border as `GLint` which must be 0 or 1](https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glTexImage2D.xml). We can probably treat it as a boolean instead of having a complicated name for the user (depends on how many bits this field is).

Reply:
Field is one bit according to the mask. I'll make it a boolean instead. Should the name be changed to reflect this?

Reply:
yeah, just `border` is fine.